### PR TITLE
webserver: add support for WSGI

### PR DESCRIPTION
**Pull request by Montellese**
This is based on #6264 (which is where the first 30 commits come from) and extends the python support in the webserver by a basic WSGI server/gateway implementation.
The following changes/additions have been made to the python codegenerator:
- support callables i.e. map C++'s `operator()` to Python C's `tp_call`
- support iterators and iterables with `%feature("iterator") <class name> "<type>"` and `%feature("iterable") <class name> "<iterator class name>" and some required implementations in C++ (`begin()`,`operator++()`,`operator==()`, ...)

The python API has been extended with a new module `xbmcwsgi` containing the classes `WsgiErrorStream`, `WsgiInputStream`, `WsgiResponse` and `WsgiResponseBody`.
Furthermore a new python invoker implementation `CHTTPPythonWsgiInvoker` has been implemented which loads the WSGI python script (defined in the webinterface's `addon.xml`) and calls the entry point function (also defined in the webinterface's `addon.xml`).

Since I have no clue about WSGI it would be nice if someone with some know-how could give this a try and see if it actually does what it should.

TODOs:
- [x] Complete documentation
- [x] Implement `exc_info` parameter in the `start_response` callable
- [x] Properly set the `PATH_INFO` environment value

### CODE REVIEW

**Comment by tamland** [comment]
@Montellese You can test wsgi compatibility by wrapping the application with the built-in validator, i.e add this to your test application:

```python
from wsgiref.validate import validator
wsgi_test = validator(wsgi_test)
````

After the `close` issue there's only one left as far as I can see: `SCRIPT_NAME` should be empty for the base path, not '/'. But I assume that's on your todo along with fixing `PATH_INFO`.

**Comment by tamland** [comment]
And I tried to run a Flask app. Didn't go so well:\ No error messages, just floods the server with request. Probably some redirect problems, possibly because `PATH_INFO` isn't set properly yet.

**Comment by Montellese** [comment]
@tamland: I'll give that a try. I guess I'll have to locally store the wsgiref implementation in the webinterface.
Yeah all those URL/path related environment variables haven't been properly implemented yet.

**Comment by tamland** [comment]
It's part of the standard library. Just add those two lines:)

**Comment by Montellese** [comment]
@tamland: Ah nice. I've added it to my `wsgi.py` but didn't see any error output (neither in the browser nor in the xbmc log).
I have adjusted the logic for `SCRIPT_NAME` to not start with a `/`. If I read PEP 333 correctly `PATH_INFO` would currently always be empty because we only forward requests to the webinterface's root path to WSGI. All other requests are handled by the webserver itself.

**Comment by da-anda** [comment]
this build is part of the experimental RPI builds from Milhouse - and with it the webinterface it no longer working. Instead it's returning some HTTP headers and other debug stuff. Just in case the debug output is/was not intended, this is the returned code: http://pastebin.com/U8V2CveR

**Comment by Montellese** [comment]
@da-anda: Yes that is the expected output that I use to test this work. Including the changes to the default webinterface (the last commit) into a test build makes no sense.

**Comment by MilhouseVH** [comment]
My bad, I thought this was getting closer to the point where it was ready for testing - I'll drop it for the time being.

**Comment by Montellese** [comment]
You can simply drop the last commit if you want to include it into your test builds. But I don't assume that anyone will suddenly start writing a WSGI based webinterface based on your test builds. But it certainly shouldn't hurt.

**Comment by MilhouseVH** [comment]
I included it mainly to determine if anything had broken, rather than for WSGI specifically. I'll try again once you think it's ready for more wide spread testing.

**Comment by Montellese** [comment]
@tamland: Could you provide a "simple" script using one of the known WSGI frameworks that I can use for further testing? Otherwise I don't really know if this is working at all or not.

We probably also have to decide how to handle requests to non-root and non-special (like `/jsonrpc`, `image`, `vfs`, `addons`, ...) paths of the webinterface. Right now they are all handled by the webserver itself. We could 
* forward them all to the WSGI application
* support a flag in the `xbmc.webinterface` extension point that would allow a webinterface author to force the webserver to forward them to the WSGI application
* forward any requests to non-existing paths to the WSGI application

**Comment by tamland** [comment]
>I've added it to my wsgi.py but didn't see any error output (neither in the browser nor in the xbmc log).

That's weird. For me it prints AssertionError *and* a nicely formated page with a stack trace.

I'll upload the example addon I have shortly;)

**Comment by tamland** [comment]
@Montellese: https://github.com/tamland/kodi-flask-demo

**Comment by tamland** [comment]
>We probably also have to decide how to handle requests to non-root and non-special (like /jsonrpc, image, vfs, addons, ...) paths of the webinterface. Right now they are all handled by the webserver itself. We could 

You forward `/addons/<addon-id>/` and subpaths to the WSGI application. Nothing else. Then *redirect* requests to `/` (only) to `/addons/<addon-id>/` where addon-id is the id for the default web interface. (Not sure if I understood you correctly here)

**Comment by Montellese** [comment]
@tamland: Rebased and tried the `wsgiref` validation again and I simply get the output that I expect.

Thanks for the demo webinterface. I had to add `markupsafe` (see https://pypi.python.org/pypi/MarkupSafe) to the addon to get it "working". Now it's constantly trying to request the root path.

**Comment by tamland** [comment]
Yeah, dumb site-packages on linux. I had it installed system wide so it silently loads that one.

**Comment by Montellese** [comment]
Thanks for the updated addon. I've temporarily added it as a submodule for easier testing. I've adjusted the logic a bit so that anything that isn't handled by any other request handler is passed to the WSGI application. That way the `/hello` request from the Flask Demo addon works fine but the root path still causes endless redirects.

**Comment by tamland** [comment]
I created another one using the 'bottle' framework too btw (https://github.com/tamland/kodi-bottle-demo). Which works fine. Both claims to be "100% WSGI compatible" so I assume there's something wrong with the environment variables causing those redirects(?)

File: xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+  environment.insert(std::make_pair("REQUEST_METHOD", requestMethod));
+
+  // SCRIPT_NAME
+  std::string path = httpRequest->path;
+  if (path.compare("/") == 0)
+    path.clear();
+  environment.insert(std::make_pair("SCRIPT_NAME", path));
+
+  // TODO: PATH_INFO
+  environment.insert(std::make_pair("PATH_INFO", path));
```
Comment (tamland) [inline xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp]: I think I know what's wrong. The spec says:
>[PATH_INFO..] may be an empty string, if the request URL targets the application root and does not have a trailing slash. 

Which implies PATH_INFO should have a trailing slash if the URL have a trailing slash.

Comment (Montellese) [inline xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp]: IMO the specification of PATH_INFO is anything but obvious. I've looked through the web and found several questions concerning it and often the answer is something like "I think [...]".

**Comment by Montellese** [comment]
OK I changed `PATH_INFO` to simply contain the whole requested path as the WSGI application will always be located at the root. That way both `/` and `/hello` work. What doesn't work yet is doing the same with `/addons/<addon-id>/` and `/addons/<addon-id>/hello`. I guess either we have to map all requests to non-`/addons/<addon-id>/` by prepending `/addons/<addon-id>/` or we have to remove `/addons/<addon-id>/` from all requests to a WSGI application.

**Comment by Montellese** [comment]
Yup, simply removing `/addons/<addon-id>` from the beginning of the requested path for WSGI webinterfaces solves that problem. A problem however might be a non-default WSGI webinterface which would generate a link which will then be missing the `/addons/<addon-id>` at the beginning.
But simply adding `/addons/<addon-id>` to all requests didn't work at all. Not sure if that would have to be defined in the WSGI application?

**Comment by tamland** [comment]
It makes a whole lot more sense if you think of paths to applications and paths withing the application. The web server should only concern itself with the `/addons/<addon-id>`part, and the application only the virtual path. In wsgi you set `SCRIPT_NAME` to `/addons/<addon-id>` and `PATH_INFO` to `/hello` etc. I think it uses that info to when generating links for instance so simply removing `SCRIPT_NAME` will cause it to link to the wrong application.

Makes to do the same elsewhere too if you ask me it. E.g for jsonrpc, instead of checking that path in `CanHandleRequest`, the webserver checks it and passes the virtual path and the path to "application" itself to the handler.

**Comment by tamland** [comment]
And can't you simply 3xx redirect `/` to `/addons/<addon-id>` and no forwarding as mentioned earlier? It means you can't use `/hello` directly, but it will be unambiguous.

**Comment by Montellese** [comment]
Added the URL redirect and fixed `SCRIPT_NAME` and `PATH_INFO`. Thanks for the explanations @tamland.

**Comment by Montellese** [comment]
I've completely rebased this and it's now completely independent from the mod_python work as this is probably the better way to go. I can still PR some of the codegenerator stuff I've added in #6264 as it might be useful in the future.

**Comment by tamland** [comment]
Session and cookies from mod_python might be useful too. Possibly those things could be provided as optional wsgi middleware? I'm not entire sure how the frameworks handle this..

**Comment by Montellese** [comment]
Thanks to @mkortstiege for the xcode sync.

@tamland: How would that be integrated into existing frameworks? I assumed that these frameworks provide session and cookie handling.

jenkins build this please

**Comment by Montellese** [comment]
Hm I don't get the errors that jenkins spits out on linux. I tried with gcc 4.8.2 and 4.9.2.
@wsnipex: Do I need to set some specific configure flags to get the same behaviour as on jenkins?

**Comment by tamland** [comment]
Try using std shared_ptr? Shouldn't use boost one anyway..

**Comment by Montellese** [comment]
@tamland: I know but I'd still like to get the same setup so that I don't run into this kind of issue again.

**Comment by Montellese** [comment]
jenkins build this please

**Comment by Montellese** [comment]
Git problems on several slaves. jenkins build this please

**Comment by Montellese** [comment]
I'm assigning this to Alpha 2 unless someone voices any complaints.

**Comment by Montellese** [comment]
jenkins build and merge

**Comment by Montellese** [comment]
OSX64 failed due to git problems and for some reason the android builds simply didn't start, see http://jenkins.kodi.tv/job/BuildMultiAndMerge-PR/177/.
jenkins build this please

**Comment by Montellese** [comment]
jenkins doesn't seem to like this PR (git fetch errors all the time) so I made a full rebase. jenkins build this please.

**Pull request merged**